base/inference.jl

```diff
@@ -862,27 +880,45 @@ function apply_type_tfunc(headtypetype::ANY, args::ANY...)
     end
     uncertain = false
     tparams = Any[]
+    outervars = Any[]
     for i = 1:largs
         ai = args[i]
         if isType(ai)
             aip1 = ai.parameters[1]
-            uncertain |= has_free_typevars(aip1)
             push!(tparams, aip1)
         elseif isa(ai, Const) && (isa(ai.val, Type) || valid_tparam(ai.val))
             push!(tparams, ai.val)
         else
             # TODO: return `Bottom` for trying to apply a non-UnionAll
-            #if !istuple && i-1 > length(headtype.parameters)
-            #    # too many parameters for type
-            #    return Bottom
-            #end
             uncertain = true
+            # These blocks improve type info but make compilation a bit slower.
+            # XXX
```

Comment:
yes. TODO is just a note for a future reader of something interesting that could be done, but isn't. But that nothing is seriously broken. XXX means something is broken or wrong. (usually very wrong, and FIXME would indicate slightly wrong, but that's pretty flexible.) Either form should also have a clarifying comment after it before ending up on master, but it's not uncommon to write one during WIP just to tag places that need revisiting before finishing the commit.